### ISSUE TITLE
Fine-granular manipulation of snapshots

### ISSUE THREAD

**Issue by ywelsch-t** [elastic/elasticsearch#11148]
###### Suggestion for more fine-granular manipulation of snapshots:

Snapshots can be partially restored by selecting only a subset of the indices. They can at the moment, however, only be fully deleted. This means that unwanted parts of a snapshot cannot be easily removed. Deleting indices from snapshots would complement the existing functionality.

Other primitive operations such as moving / copying indices between snapshots might be useful. At the moment, only a single snapshot operation can be active in the cluster. This determines the level of concurrency for backing up as well as restoring data. In a large cluster, the backing up of data might be partitioned into smaller snapshot tasks. This is relevant when prioritizing certain indices or backing up indices with different frequencies. It also helps to minimize the interference with re-balancing processes by backing up individual indices as fast as possible (reduces the amount of transfer time for each individual index). Now if I snapshot the first ten indices and then ten more (e.g. to consider priorities), I will not be able to restore index A of the former snapshot and index B of the latter snapshot at the same time. Moving indices between snapshots would allow the combination of both snapshots.
###### What about immutability of snapshots?

At the moment, snapshots are immutable (they are created as a fixed set of indices). Adding delete / move / copy operations would change that. However, other operations could be imagined that conform to the "immutable" model. For example, creating a new snapshot based on the content of other snapshots (by giving an ordered list of (snapshot, indices) pairs). With the current implementation of snapshots, this would be very fast (only writing meta-data).

I can make a pull request if this is a feature that others deem useful. Also, the specifics of the operations need to be clarified first.

**Comment by imotov** [comment]
I think we can keep the immutability property by creating a new snapshot from pieces of existing snapshots, so instead of move/copy you can have functionally equivalent partial merge. However, currently a snapshot represents a point in time state of a cluster (or part of a cluster) at the moment the snapshot was created. It's nice property to have for several reasons. The most important reason is knowing the version of the cluster that wrote the snapshot and therefore being able to make different discussions based on this version. Slicing and gluing pieces of different snapshots together removes this property and makes handling older versions of the snapshots more complicated. So, considering the additional complexity that this feature will introduce I am not sure if the use case of restoring data from two snapshots at the same time is worth it. @ywelsch-t, could you elaborate a bit more on the benefits of combining two separate snapshots into one comparing to restoring them one after another.

**Comment by ywelsch-t** [comment]
Let me try and explain some more. The setting we have is backing up data using the Hadoop HDFS Snapshot/Restore plugin. Source of the backup is a large ES cluster (> 500 nodes) with lots of data and destination is an HDFS cluster (many nodes as well).

The tricky part is finding the right amount of throughput between the ES cluster and the HDFS cluster such that
- our indices that we want to backup / restore are transfered as fast as possible but
- without impacting the performance of the cluster too much (i.e. saturating the shared network for regular ES communication).

The HDFS cluster is specifically set up to store ES backups and has no resource limitations.

Currently, the only way to control speed and concurrency with ES is using three parameters:
- max_snapshot_bytes_per_sec / max_restore_bytes_per_sec which trottles the bandwidth per node.
- The parameter concurrent_streams for the HDFS plugin determines the amount of parallel transfers of low-level files (e.g. Lucene segments) per node.
- Amount of nodes participating in the snapshot / restore process. This cannot be directly selected but is determined 
  - by the number of shards that are part of the snapshot and
  - by the distribution of shards in the cluster (i.e. allocation to node)

A limitation which we face is that multiple snapshots / restore actions cannot run concurrently.

This means that:
- To reach a certain throughput, we must snapshot / restore as many shards as possible in parallel.
- To not oversaturate our network, we cannot have too many nodes (and hence shards) as part of the backup action. This means that only max. 10% of the nodes in our cluster should be involved at the same time in backup operations.

Our solution is thus to backup batches of similarly sized indices (and hence shards) to have constant throughput.

In conclusion, our snapshots are sized based on transfer capabilites and network concerns.

However, this leads to the following conflicts if a snapshot can only be manipulated / deleted in its whole:
1) Some indices (in a snapshot) have different life spans than others. The operation "Delete index from snapshot" is useful in that case to "clean up" older snapshots. Deleting some of the indices in a snapshot also does not violate the property that a snapshot represents a point in time of a cluster (or part of a cluster) at the moment the snapshot was created.

2) In case of an emergency restore, we want to restore as fast as possible (i.e. fully saturate the network). Additionally, we need to prioritize the restore process of the most relevant indices. With the possibility to create a new snapshot based on the contents of other snapshots (or to merge snapshots), we can selectively pick the indices from different snapshots that we want to restore, create fake snapshots based on them and restore the fake snapshots in custom order.


We have patched ES with an implementation to create fake snapshots based on the contents of other snapshots and to delete parts of a snapshot. We believe this fine-granular snapshot manipulation gives us the flexibility for more customized backup strategies.

**Comment by imotov** [comment]
@ywelsch-t this is a great explanation. Thank you! So, if we assume for a moment that besides node-level throttling you have a cluster-level (or better to say repository-level) throttling for both snapshot and restore operations, would it make this feature unnecessary?

**Comment by ywelsch-t** [comment]
Repository-level throttling is only part of the solution. It solves the issue of snapshot transfer impacting the overall performance of the cluster too much (i.e. saturating the shared network for regular ES communication).

Two issues I have described in the previous comment are remaining, however.

1. To reach a certain throughput, we must snapshot / restore multiple indices in parallel.
  - For backups, this means that we may have to combine different indices into a single snapshot although they have different lifetimes.
  - For restores, this means that we have to combine the restore of indices in different snapshots.

2. Prioritization. This is particularly important for restore operations (see point 2 of previous post).


For SNAPSHOT operations, I see three solutions:
a) allow concurrent snapshots (of disjoint indices sets). We then group indices by lifetimes / prioritization and do concurrent backups.
b) give possibility to dissect snapshot after the backup. The initial snapshot is then only a means of concurrent transfer. If lifetime expires, we delete the "expired" indices from the snapshot. For that we need an operation "delete index from snapshot" (which, as previously said, does not conflict with the original intent of snapshots). Advantage is that lifetime of certain indices in the backup can be controlled not only before but also after snapshotting.
c) allow both a) and b). My preferred solution.

For RESTORE operations, I see two solutions:
a) allow concurrent snapshot restores (of disjoint indices sets). My preferred solution.
b) give possibility to combine snapshots before the restore. This is a poor substitute for a).

**Comment by imotov** [comment]
@ywelsch-t ok, so if we will enable concurrent snapshots of non-intersecting sets of indices into different repositories and concurrent restore plus repository-level throttling, will it completely solve the issue?

**Comment by ywelsch-t** [comment]
1) Why the restriction to different repositories? If snapshots are concurrently done on the same repository but for disjoint sets of indices, isn't it only the update of the snapshot list at the end that needs to be synchronized (as far as BlobStoreRepository is concerned)?

2) I would prefer to also have official support for the "delete index from snapshot" operation (Advantage is that lifetime of certain indices in the backup can be controlled not only before but also after snapshotting.) . A workaround would be to do snapshots that are only comprised of single indices.

3) Repository-level throttling becomes actually less important if we have concurrent snapshots/restores because we can then approximately control the level of network throughput by sizing the snapshot / restore operations appropriately. Also, I am not sure whether throttling at the repository level makes the most sense. Should we do it at a cluster level, or even at a per-snapshot or per-restore level? How does it relate to node-level throttling? Could nodes as well be throttled at a per-snapshot or per-restore level? For me, this feature has the most unknowns right now.

In conclusion, what would solve our issues:
- concurrent snapshots of disjoint sets of indices
- concurrent restore of disjoint sets of indices from different snapshots
- "delete index from snapshot" operation

A nice-to-have:
- some form of throttling for snapshot and restore operations.

**Comment by ywelsch-t** [comment]
@imotov I could help contribute on the following items:
- concurrent snapshots of disjoint sets of indices
- concurrent restore of disjoint sets of indices from different snapshots
- "delete index from snapshot" operation
Is that something you would consider accepting as PR?

**Comment by imotov** [comment]
@ywelsch-t concurrent snapshot and concurrent restore is definitely on the roadmap. The infrastructure is there that can be enabled with a couple of lines of code, but it is lacking good test coverage. So, you can definitely start working on that if you wish. 

I am not completely sold on the "delete index from snapshot" operation to be honest. I would wait for others to comment.

**Comment by tlrx** [comment]
My current understanding is that we need a) concurrent snapshot and concurrent restore of non-intersecting sets of indices b) throttling at repository level in order to not saturate the repository (ie limit the number of shards that are concurrently snapshot to/restored from the repo) c) throttling at cluster level in order to limit impact on cluster performance.

With concurrent restore and the fact it’s already possible to select indices that should be restored by using `indices` I'm not sure we need a "delete index from snapshot" operation. The only advantage I see is to clean and reduce a too large snapshot to free space. If indices have different lifespans, maybe they should be put into different snapshots?

**Comment by 4ydx** [comment]
Just to speak up concurrent restoration of non-intersecting sets of indices would be much appreciated.  I am not at a familiar with the internals of ES, but if somebody were willing to point me in the right direction, I would be happy to take a look at how that could be enabled in the code.  My company needs this feature in the coming months.

**Comment by 4ydx** [comment]
I was wondering if somebody might be able to take a look at

https://github.com/4ydx/elasticsearch/commit/e10ac70ef7c5f0bc1d06478dd9efda94fcf85cd5

and let me know if those minor changes seem reasonable for allowing concurrent restoration.  I am currently running mvn clean test as per the "Contributing to Elasticsearch" docs.  As a side note an unaltered clone followed by running mv clean test is producing failures.  I submitted another issue to find out if this is common or not.  At the end of the day I would like to be able to contribute something if possible, but please remember that I am new to Java as well as the Elasticsearch codebase.

**Comment by clintongormley** [comment]
Hi @4ydx 

Best idea is to send your changes as a pull request - here it'll just get lost.

**Comment by balooka** [comment]
I'll add my vote regarding the "delete indice from snapshot" feature,
As it'll definitely allow for cleanup in 2 occasions I've encountered in my current environment:
- as described by @ywelsch-t, allowing better management of indices with different life spans, although in our case the reason for snapshots containing such a mix relates more to original strategy vs actual need. Reviewing and amending the snapshot strategy,  I'm still "locked" by the older snapshots created, where fairly small indices that should be kept are mixed with larger indices that consume all the spaces but could/should be deleted because there is no need to keep them any longer
- due to filesystem space issue, some snapshots have only completed partially, so we wouldn't be able to use safely the snapshot for some of the indices anyway, better to be able to clean them then

Another approach that would be completely satisfactory to our use case (but doesn't cover the throughput aspect brought up in this discussion) would be to allow for snapshot generations based on existing snapshots, only allowing splitting (new snapshot has a subset of the indices of the older snapshot)

**Comment by DravenJohnson** [comment]
1. How about add combine/merge feature that you can organize your snapshot every month (daily snapshot merge into a monthly one.)

2. I would prefer if we have retake option for failed snapshot so I don't have to spend more time retaking entire snapshot. (Time consuming and also not well organized.)

**Comment by ywelsch** [comment]
I'm closing this issue as we're going to tackle the problem from a different angle. In the future, we plan on having a more index-centric view of snapshots. Instead of asking what indices are in a snapshot, you could ask what snapshots exist for an index and do manipulation at index level. This would address the issues mentioned here.

**Comment by jakauppila** [comment]
@ywelsch Is there another issue to watch for that index-centric view of snapshots? We'd be very interested in that.
